### PR TITLE
Set qubes providing input devices to remain powered on

### PR DESCRIPTION
**Pull request by fepitre**
https://github.com/QubesOS/qubes-issues/issues/9034

### CODE REVIEW

File: qubes-dist-upgrade.spec.in [hunk 1]
```diff
@@ -6,6 +6,8 @@ License:	GPLv2
 BuildArch:	noarch
 Source0:	%{name}-%{version}.tar.gz
 
+BuildDepends: xinput
```
Comment (marmarek) [inline qubes-dist-upgrade.spec.in]:
```suggestion
Requires: xinput
```

File: qubes-dist-upgrade.sh [hunk 2]
```diff
@@ -212,6 +213,20 @@ max_concurrency="${max_concurrency:-4}"
...
 update_prechecks
 
+# Automatically add qubes providing dom0 input devices to remain powered on
+if which xinput > /dev/null; then
+    read -r -a xinput_qubes <<< "$(xinput list --name-only | sed -n 's/\(^[a-zA-Z][a-zA-Z0-9_-]*\):.*$/\1/p')"
+    for qube in "${xinput_qubes[@]}"
+    do
+        if qvm-check -q "$qube"; then
+            echo "INFO: Found $qube providing input devices. It will be kept running."
+            extra_keep_running+=( "$qube" )
```
Comment (marmarek) [inline qubes-dist-upgrade.sh:224]: This (currently) won't be enough to keep it running in stage 3. See the first task in the ticket description.

And also, after stage 3 no new qrexec service will work, which means if somebody unplug USB mouse/keyboard, plugging it back wont work. Maybe include it in the warning too?

**Review by marmarek — CHANGES_REQUESTED** [review]

File: qubes-dist-upgrade.spec.in (hunk 1 above)
Comment (fepitre) [reply qubes-dist-upgrade.spec.in]: Yes stupid..

File: qubes-dist-upgrade.sh [hunk 3]
```diff
@@ -327,8 +327,12 @@ if [ "$assumeyes" == "1" ] || confirm "-> Launch upgrade process?"; then
...
         qubes-dom0-update ${dnf_opts_noclean} --downloadonly --force-xen-upgrade --action=distro-sync || exit_code=$?
         if [ -z "$exit_code" ] || [ "$exit_code" == 100 ]; then
-            if [ "$assumeyes" == "1" ] || confirm "---> Shutdown all VM?"; then
-                qvm-shutdown --wait --all
+            if [ "$assumeyes" == "1" ] || confirm "---> Shutdown all qubes (requested keep running and input devices qubes will remain powered on)?"; then
+                qvm_shutdown_exclude=()
+                for qube in "${extra_keep_running[@]}"; do
+                    qvm_shutdown_exclude+=("--exclude $qube")
+                done
+                qvm-shutdown --wait --all "${qvm_shutdown_exclude[@]}"
```
Comment (fepitre) [inline qubes-dist-upgrade.sh:336]: @marmarek it should be what's needed for keeping running on stage 3?

Comment (marmarek) [reply qubes-dist-upgrade.sh:336]: Should be okay I think

File: qubes-dist-upgrade.sh [hunk 4]
```diff
@@ -142,6 +139,14 @@ restore_rpmsave_policy() {
     done
 }
 
+has_updated_qubes_version() {
+    if grep -q 'VERSION_ID=4.2' /etc/os-release; then
+        true
+    else
+        false
+    fi
```
Comment (marmarek) [inline qubes-dist-upgrade.sh]: just the `grep -q` line? the if+true+false part is redundant

**Review by marmarek — CHANGES_REQUESTED** [review]

File: qubes-dist-upgrade.sh (hunk 4 above)
Comment (fepitre) [reply qubes-dist-upgrade.sh]: yes, it was some pending code to know if more checks were needed or not like checking also xen version maybe?

Comment (fepitre) [reply qubes-dist-upgrade.sh]: In all the cases, it can be a one line indeed.

File: qubes-dist-upgrade.sh [hunk 5]
```diff
@@ -312,8 +328,12 @@ if [ "$assumeyes" == "1" ] || confirm "-> Launch upgrade process?"; then
         # shellcheck disable=SC2086
         qubes-dom0-update ${dnf_opts_noclean} --downloadonly --force-xen-upgrade --action=distro-sync || exit_code=$?
         if [ -z "$exit_code" ] || [ "$exit_code" == 100 ]; then
-            if [ "$assumeyes" == "1" ] || confirm "---> Shutdown all VM?"; then
-                qvm-shutdown --wait --all
+            if [ "$assumeyes" == "1" ] || confirm "---> Shutdown all qubes (requested keep running and input devices qubes will remain powered on)?"; then
+                qvm_shutdown_exclude=()
+                for qube in "${extra_keep_running[@]}"; do
+                    qvm_shutdown_exclude+=("--exclude $qube")
```
Comment (marmarek) [inline qubes-dist-upgrade.sh]: This doesn't work: https://openqa.qubes-os.org/tests/99213#step/release_upgrade/29
I think you need:
```suggestion
                    qvm_shutdown_exclude+=("--exclude" "$qube")
```
or maybe with `=` ?

**Review by marmarek — CHANGES_REQUESTED** [review]

File: qubes-dist-upgrade.sh (hunk 5 above)
Comment (fepitre) [reply qubes-dist-upgrade.sh]: It works with `=`.

**Review by marmarek — APPROVED** [review]

**Pull request merged**